pctasks_funcs/StorageEventsCF/__init__.py

```diff
@@ -0,0 +1,118 @@
+CosmosDB to the appropriate Blob Storage Queue. It's expected that PCTasks
+workflows will scale in response to messages being sent to that queue.
+"""
+import contextlib
+import logging
+import os
+import re
+from typing import Optional
+
+import azure.functions as func
+import azure.identity.aio
+import azure.storage.queue.aio
+
+from pctasks.core.models.event import StorageEvent
+
+# TODO: pre-commit-style validator for ensuring these queues exist.
+# TODO: Move to prefix / suffix rules?
+QUEUE_ENDPOINT = "https://pclowlatency.queue.core.windows.net/"
+
+RULES = [
+    # GOES-GLM
+    (
+        re.compile(
+            r"https://goeseuwest\.blob\.core\.windows\.net\/noaa-goes(16|17|18)/GLM-L2-LCFA/"  # noqa: E501
+        ),
+        "goes-glm",
+    ),
+    # GOES-CMI (or at least other)
+    (
+        re.compile(
+            r"https://goeseuwest\.blob\.core\.windows\.net\/noaa-goes(16|17|18)/"
+        ),
+        "goes-cmi",
+    ),
+    # sentinel-1-grd
+    (
+        re.compile(
+            r"https://sentinel1euwest\.blob\.core\.windows\.net/s1-grd-stac/GRD/"
+        ),
+        # technically could do pre-made items queue
+        "sentinel-1-grd",
+    ),
+    # sentinel-1-rtc
+    (
+        re.compile(
+            r"https://sentinel1euwestrtc\.blob\.core\.windows\.net/sentinel1-grd-rtc-stac/GRD/"  # noqa: E501
+        ),
+        # technically could do pre-made items queue
+        "sentinel-1-rtc",
+    ),
+    # For integration tests. It doesn't feel great putting this in here.
+    (re.compile(r"http://azurite:10000/devstoreaccount1/"), "test-collection"),
+]
+
+
+def dispatch(url: str) -> Optional[str]:
+    """
+    Parameters
+    ----------
+    """
+    # TODO: Decide whether we want 1:1 or 1:many
+    for rule, queue_name in RULES:
+        if rule.match(url):
+            logging.info("message=matched, rule=%s, queue-name=%s", rule, queue_name)
+            return queue_name
+    return None
+
+
+async def main(documents: func.DocumentList) -> None:
+    # Types of auth to support
+    # 1. Connection string (azurite)
+    # 2. DefaultAzureCredential (prod)
+    # TODO: local.settings.json
+    account_url = os.environ["FUNC_STORAGE_QUEUE_ACCOUNT_URL"]
+    credential = os.environ.get("FUNC_STORAGE_ACCOUNT_KEY", None)
+
+    if credential is None:
+        credential = azure.identity.aio.DefaultAzureCredential()  # type: ignore
+        credential_ctx = credential
+    else:
+        credential = {  # type: ignore
+            "account_name": os.environ["FUNC_STORAGE_ACCOUNT_NAME"],
+            "account_key": credential,
+        }
+        # async support for contextlib.nullcontext is new in 3.10
+        # credential_ctx = contextlib.nullcontext
+
+        @contextlib.asynccontextmanager
+        async def credential_ctx_():  # type: ignore
+            yield
+
+        credential_ctx = credential_ctx_()
+
+    async with credential_ctx:  # type: ignore
```

Comment:
This logic looks similar to the other changefeed function; could this logic be consolidated?

Reply:
I've actually never tried to share code between Azure functions. https://learn.microsoft.com/en-us/azure/azure-functions/functions-reference-python?pivots=python-mode-configuration&tabs=asgi%2Capplication-level#import-behavior indicates we should be able to put a package at the top level and import it in both places. I'll give that a shot.

Reply:
As an example, there used to be a "func_lib" package in the functions folder that was used across functions: https://github.com/microsoft/planetary-computer-tasks/tree/c961586b4346f055e7dcf930c12b65816ef66724/pctasks_funcs